fix: Avoid fatal merge-base calls on shallow checkouts

nmote opened the pull request:
#6756 allows us to use the GitHub API to find the merge base, rather than relying on the sometimes slow process of fetching additional commits until we can compute it locally.

However, downstream code assumes that we have enough history to compute the merge base. When it tries to do so, Git fatals and Semgrep aborts. This was reported to us when a customer tried to make use of this optimization.

I missed this while testing the original PR because my test PR on my test repo added the file with the finding. When all the files with findings are added in the PR being scanned, Semgrep doesn't bother doing a baseline scan, which I overlooked.

Test plan:

Same as #6756 except the test PR modifies an existing file with findings rather than creating a new file. When using the current `returntocorp/semgrep` image, Semgrep fails on the merge base call in the same way as described by the customer. When using the locally-built `nmote/semgrep` image with the change in this PR, Semgrep runs correctly.

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

nmote commented:
I don't think I want to address the Semgrep finding. All I did was modify an existing string that is logged.

aryx commented on cli/src/semgrep/git.py:
```diff
@@ -338,10 +341,10 @@ def baseline_context(self) -> Iterator[None]:
-            ["git", "merge-base", self._base_commit, "HEAD"]
+        merge_base_sha = self._get_git_merge_base()
+        logger.info(
+            "  Will report findings introduced by these commits (may be incomplete for shallow checkouts):"
```
is it possible to detect shallow checkout so that we print the additional warning only if we know it's a shallow checkout?

nmote replied:
I'm not sure. At this point I'm reluctant to complicate this further. Do you think it's critical?

nmote replied:
I'm going to merge, but am happy to follow up if you think it's important.

emjin commented on cli/src/semgrep/git.py:
```diff
@@ -196,6 +196,12 @@ def _get_git_status(self) -> GitStatus:
         return GitStatus(added, modified, removed, unmerged, renamed)
 
+    def _get_git_merge_base(self) -> str:
+        if self._is_mergebase:
```
Can you explain why we need this check? Wasn't obvious to me so it would be nice to preserve it

nmote committed:
fix: Avoid fatal merge-base calls on shallow checkouts

#6756 allows us to use the GitHub API to find the merge base, rather
than relying on the sometimes slow process of fetching additional
commits until we can compute it locally.

However, downstream code assumes that we have enough history to compute
the merge base. When it tries to do so, Git fatals and Semgrep aborts.
This was reported to us when a customer tried to make use of this
optimization.

I missed this while testing the original PR because my test PR on my
test repo added the file with the finding. When all the files with
findings are added in the PR being scanned, Semgrep doesn't bother doing
a baseline scan, which I overlooked.

Test plan:

Same as #6756 except the test PR modifies an existing file with findings
rather than creating a new file. When using the current
`returntocorp/semgrep` image, Semgrep fails on the merge base call in
the same way as described by the customer. When using the locally-built
`nmote/semgrep` image with the change in this PR, Semgrep runs
correctly.
